internal/services/iothub/iothub_resource.go

```diff
@@ -1269,14 +1277,20 @@ func expandIoTHubEndpoints(d *pluginsdk.ResourceData, subscriptionId string) (*d
 	eventHubProperties := make([]devices.RoutingEventHubProperties, 0)
 	storageContainerProperties := make([]devices.RoutingStorageContainerProperties, 0)
 
-	for _, endpointRaw := range routeEndpointList {
+	for k, endpointRaw := range routeEndpointList {
 		endpoint := endpointRaw.(map[string]interface{})
 
 		t := endpoint["type"]
 		name := endpoint["name"].(string)
 		resourceGroup := endpoint["resource_group_name"].(string)
 		authenticationType := devices.AuthenticationType(endpoint["authentication_type"].(string))
-		subscriptionID := subscriptionId
+
+		subscriptionID := endpoint["subscription_id"].(string)
+		// We have to set `subscription_id` with the subscription Id in the current context when `subscription_id` isn't specified in the tf config, otherwise it would block the existing users
+		// As `Computed: true` is enabled, TF would always get the value from the last apply when this property isn't set. So we have to use d.GetRawConfig() to determine if it's set in the tf config
+		if v := d.GetRawConfig().AsValueMap()["endpoint"].AsValueSlice()[k].AsValueMap()["subscription_id"]; v.IsNull() {
+			subscriptionID = subscriptionId
+		}
```

Comment:
Instead of setting the schema as O+C, can we make it Optional only by adding a restriction to this attribute: If the subscription id is the same as the provider id, simply set it as null or an empty string. This makes sense to me as users need to specify this attribute, only when it is different than the provider's subscription id.

With this, you only need to do a conditional set to state during the flattening.

Reply:
If we do it this way, it will result in the `subscription_id` on the service side having a value, but the `subscription_id` in the state file will be empty. This might confuse the user, so I still recommend using my method to fix it.

Reply:
The current implementation has no issue in itself, but I'm more concerned about the complexity it brings, from the maintenance perspective. If we decide to go down this route, please remember to add comments to clearly describe the raw config handling is only deant for the case that the user has a eventhub whose endpoint's subscription is not the provider's one. Then the user wants to reset it to the provider's one by unset the `subscription_id`. The current comment doesn't make sense to me.

Reply:
Sure. I updated the descriptions.

Reply:
Do we really need to do this? and how reliable is it given the issues we have with GetOkExists etc not working in every situation?

i do worry about adding this as it will be technical debt for when the transition to framework is done 

this is only to enable a user to unset the value and have it default to the current provider subscription?

Reply:
Yes. Using "d.GetRawConfig()" is to align with the previous TF behavior to enabled a user to unset the value and have it default to the current provider subscription for unblocking the existing users.